src/test/PulsarMock.js

```diff
@@ -0,0 +1,229 @@
+export class PulsarClientMock {
+  static inMemoryStore = new Map();
+  static messageIdCounter = 0;
+  static listeners = new Map();
+  static ackTimeout = 30000;
+  static acknowledgedMessages = new Map();
+  static operationLocks = new Map(); // Map to keep track of ongoing operations by topic
+
+  // Helper method to lock operations per topic
+  static async lockOperation(topic) {
+    while (this.operationLocks.get(topic)) {
+      await this.operationLocks.get(topic);
+    }
+    let resolveLock;
+    const lockPromise = new Promise(resolve => resolveLock = resolve);
+    this.operationLocks.set(topic, lockPromise);
+    return resolveLock;
+  }
+
+  static configureAckTimeout(timeout) {
+    this.ackTimeout = timeout;
+  }
+
+  async createProducer(config) {
+    const topic = config.topic;
+    if (!PulsarClientMock.inMemoryStore.has(topic)) {
+      PulsarClientMock.inMemoryStore.set(topic, []);
+    }
+
+    return {
+      send: async (message) => {
+        const resolveLock = await PulsarClientMock.lockOperation(topic); // Lock operation for topic
+        try {
+          const messageId = new MessageId(`message-${++PulsarClientMock.messageIdCounter}`);
+          const messageInstance = new Message(topic, undefined, message.data, messageId, Date.now(), Date.now(), 0, '');
+          const messages = PulsarClientMock.inMemoryStore.get(topic);
+          messages.push({ message: messageInstance, visible: true });
+
+          PulsarClientMock.notifyListeners(topic);
+          return { messageId: messageId.toString() };
+        } finally {
+          resolveLock(); // Unlock operation for topic
+          PulsarClientMock.operationLocks.delete(topic); // Cleanup lock
+        }
+      },
+      close: async () => {},
+    };
+  }
+
+  async subscribe(config) {
+    const topic = config.topic;
+
+    return {
+      receive: () => {
+        return new Promise((resolve) => {
+          const tryResolve = () => {
+            const messages = PulsarClientMock.inMemoryStore.get(topic) || [];
+            const messageIndex = messages.findIndex(m => m.visible);
+            if (messageIndex !== -1) {
+              // Make the message temporarily invisible to simulate message locking
+              messages[messageIndex].visible = false;
+              setTimeout(() => {
+                // Make the message visible again after the timeout
+                messages[messageIndex].visible = true;
+                PulsarClientMock.notifyListeners(topic);
+              }, PulsarClientMock.ackTimeout);
+              resolve(messages[messageIndex].message);
+            } else {
+              // No visible messages available, wait for new messages
+              if (!PulsarClientMock.listeners.has(topic)) {
+                PulsarClientMock.listeners.set(topic, []);
+              }
+              PulsarClientMock.listeners.get(topic).push(tryResolve);
```

Comment:
I have never code that tries to utilize a function from inside the definition of the function. Super weird. Can you write this in a more idiomatic way?

Reply:
The idea with this code that a listener is each `receive` invocation, which was blocked because there is no new messages. What I realized from looking at this code, that we can't have more than one subscriber for this model, so I extended it to support per-subscriberId support.